js/src/path.js

```diff
@@ -4,43 +4,220 @@
 // Licensed under the Apache License, version 2.0:
 // http://www.apache.org/licenses/LICENSE-2.0
 
+import {invariant} from './assert.js';
 import type Value from './value.js';
+import Hash from './hash.js';
 import {Kind} from './noms-kind.js';
 import List from './list.js';
 import Map from './map.js';
+import Set from './set.js';
+import {OrderedKey} from './meta-sequence.js';
+import {OrderedSequence} from './ordered-sequence.js';
+import {fieldNameComponentRe} from './struct.js';
 import {getTypeOfValue, StructDesc} from './type.js';
 
-interface Part {
-  resolve(v: Promise<?Value>): Promise<?Value>;
+const annotationRe = /^@([a-z]+)/;
+
+export interface Part {
+  resolve(v: Value): Promise<?Value>;
   toString(): string;
 }
 
-class FieldPart {
-  name: string;
+/**
+ * A Path is an address to a Noms value - and unlike hashes (i.e. #abcd...) they can address inlined
+ * values. See https://github.com/attic-labs/noms/blob/master/doc/spelling.md.
+ */
+export default class Path {
+  _parts: Array<Part>;
 
-  constructor(name: string) {
-    this.name = name;
+  static parse(str: string): Path {
```

Comment:
I would prefer this to be an exported function of this module instead of a static method of the class... or make it work with sub classing...